src/app/icd/server/ICDManager.cpp

```diff
@@ -234,15 +240,87 @@ bool ICDManager::CheckInMessagesWouldBeSent()
+ * having a single implementation of the function renders the implementation very difficult to understand and maintain.
+ * Because of this, each valid feature combination has its own implementation of the function.
+ */
+#if CHIP_CONFIG_PERSIST_SUBSCRIPTIONS
+#if CHIP_CONFIG_SUBSCRIPTION_TIMEOUT_RESUMPTION
+/**
+ * @brief Implementation for when the persistant subscription and subscription timeout resumption feature are present.
+ *        Function checks that there no active or persisted subscriptions for a given fabricIndex or subjectID.
+ *
+ * @param aFabricIndex
+ * @param subjectID subjectID to check. Can be an operational node id or a CAT
+ *
+ * @return true Returns true if the fabricIndex and subjectId combination does not have an active or a persisted subscription.
+ * @return false Returns false if the fabricIndex and subjectId combination has an active or persisted subscription.
+ */
+bool ICDManager::ShouldCheckInMsgsBeSentAtActiveModeFunction(FabricIndex aFabricIndex, NodeId subjectID)
+{
+    return !(mSubInfoProvider->SubjectHasActiveSubscription(aFabricIndex, subjectID) ||
+             mSubInfoProvider->SubjectHasPersistedSubscription(aFabricIndex, subjectID));
+}
+#else
+/**
+ * @brief Implementation for when the persistant subscription feature is present without the subscription timeout resumption
+ * feature. Function checks that there no active subscriptions. If the boot up subscription resumption has not been completed,
+ *        function also checks if there are persisted subscriptions.
+ *
+ * @note The persistent subscriptions feature tries to resume subscriptions at the highest min interval
+ *       of all the persisted subscriptions. As such, it is possible for the ICD to return to Idle Mode
+ *       until the timer elaspses. We do not want to send Check-In messages to clients with persisted subscriptions
+ *       until we have tried to resubscribe.
+ *
+ * @param aFabricIndex
+ * @param subjectID subjectID to check. Can be an opperationnal node id or a CAT
+ *
+ * @return true Returns true if the fabricIndex and subjectId combination does not have an active subscription.
+ *              If the boot up susbscription has not been completed, there must not be a persisted subscription either.
+ * @return false Returns false if the fabricIndex and subjectId combination has an active subscription.
+ *               If the boot up susbscription has not been completed,
+ *               returns false if the fabricIndex and subjectId combination has a persisted subscription.
+ */
+bool ICDManager::ShouldCheckInMsgsBeSentAtActiveModeFunction(FabricIndex aFabricIndex, NodeId subjectID)
+{
+    bool wouldSendCheckIn = !(mSubInfoProvider->SubjectHasActiveSubscription(aFabricIndex, subjectID));
+
+    if (!mIsBootUpResumeSubscriptionExecuted)
+    {
+        wouldSendCheckIn = wouldSendCheckIn && !mSubInfoProvider->SubjectHasPersistedSubscription(aFabricIndex, subjectID);
+    }
+
+    return wouldSendCheckIn;
```

Comment:
This might be more readable as:
```suggestion
    bool mightHaveSubscription = mSubInfoProvider->SubjectHasActiveSubscription(aFabricIndex, subjectID);
    if (!mightHaveSubscription && !mIsBootUpResumeSubscriptionExecuted) {
        mightHaveSubscription = mSubInfoProvider->SubjectHasPersistedSubscription(aFabricIndex, subjectID);
    }
    
    return !mightHaveSubscription;
```

but I have to admit that's largely a matter of taste, so don't feel too strongly about it.